performance/vip-tweaks.php

```diff
@@ -25,50 +39,118 @@ function wpcom_vip_disable_performance_tweaks() {
-	$cached_latest_month = ( ! empty( $media_months[0]->month ) ) ? $media_months[0]->month : '';
+	return wpcom_vip_get_available_months_for_filters( $post_type );
+}
 
-	// If the transient exists, and the attachment uploaded doesn't match the first (latest) month or year in the transient, lets clear it.
-	$matches_latest_year  = get_the_time( 'Y', $post_id ) === $cached_latest_year;
-	$matches_latest_month = get_the_time( 'n', $post_id ) === $cached_latest_month;
-	if ( false !== $media_months && ( ! $matches_latest_year || ! $matches_latest_month ) ) {
-		// the new attachment is not in the same month/year as the data in our transient
-		delete_transient( 'wpcom_media_months_array' );
+/**
+ * Query wrapper that caches the results of available months given a particular post type.
+ *
+ * See WP_List_Table::months_dropdown() & wp_enqueue_media() in WP core.
+ *
+ * @param string $post_type
+ * @return object[]
+ */
+function wpcom_vip_get_available_months_for_filters( $post_type ) {
+	global $wpdb;
+
+	$cache_key = "available_filter_months_$post_type";
+	$months    = wp_cache_get( $cache_key, 'vip' );
+	if ( is_array( $months ) ) {
+		// Happiest-path, cache exists already :).
+		return $months;
 	}
+
+	$extra_checks = '';
+	if ( 'attachment' !== $post_type ) {
+		$extra_checks = " AND post_status != 'auto-draft' AND post_status != 'trash'";
+	}
+
+	// phpcs:disable WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+	$months = $wpdb->get_results(
+		$wpdb->prepare(
+			"
+			SELECT DISTINCT YEAR( post_date ) AS year, MONTH( post_date ) AS month
+			FROM $wpdb->posts
+			WHERE post_type = %s
+			$extra_checks
+			ORDER BY post_date DESC
+			",
+			$post_type
+		)
+	);
+	// phpcs:enable WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+
+	wp_cache_set( $cache_key, $months, 'vip', DAY_IN_SECONDS );
+	return $months;
 }
 
-if ( is_admin() ) {
-	add_filter( 'media_library_show_video_playlist', '__return_true' );
-	add_filter( 'media_library_show_audio_playlist', '__return_true' );
-
-	add_filter( 'media_library_months_with_files', 'wpcom_vip_media_library_months_with_files' );
-	function wpcom_vip_media_library_months_with_files() {
-		global $wpdb;
-
-		$months = get_transient( 'wpcom_media_months_array' );
-
-		if ( false === $months ) {
-			// phpcs:ignore WordPress.DB.DirectDatabaseQuery
-			$months = $wpdb->get_results( $wpdb->prepare( "
-				SELECT DISTINCT YEAR( post_date ) AS year, MONTH( post_date ) AS month
-					FROM $wpdb->posts
-					WHERE post_type = %s
-					ORDER BY post_date DESC
-				", 'attachment' ) );
-			set_transient( 'wpcom_media_months_array', $months );
+/**
+ * Helper method to bust the "available months" cache when needed.
+ *
+ * Note that we only bust the cache when a new month is added,
+ * For deletions, we just let the 24-hour TTL handle that.
```

Comment:
Why did you choose this approach over having no cache expiry and invalidating cache on post deletion?

Reply:
Most post deletions are quite unlikely to result in any change to the "available months". It would have to be the last post within that month to result in a change. Worst case is that the "last post" is deleted and somebody filters by a month within the same day and finds no posts there, but truly unlikely. Vs the high likelihood of deletes being fairly often and resulting in cache busts.